### PR TITLE
feat: implement ADR-20: Post ownership transfer

### PR DESCRIPTION
**Pull request by dadamu**
## Description

This PR implements ADR-20: Post ownership transfer.

Closes: [DCD-320](https://desmoslabs.atlassian.net/browse/DCD-320)

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/desmos-labs/desmos/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://docs.cosmos.network/v0.44/building-modules/intro.html)
- [x] included the necessary unit and integration [tests](https://github.com/desmos-labs/desmos/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

[DCD-320]: https://forbole.atlassian.net/browse/DCD-320?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/architecture/adr-020-post-ownership-change.md [hunk 1]
```diff
@@ -68,10 +68,10 @@ message PostOwnerTransferRequest {
 To simplify the management of post owner transfer requests, we will store each request using the following key format:
 
 ```
-IncomingPostOwnerTransferRequestPrefix | SubspaceID | ReceiverAddress | PostID | -> Protobuf(PostOwnerTransferRequest)
+PostOwnerTransferRequestPrefix | SubspaceID | PostID | -> Protobuf(PostOwnerTransferRequest)
 ```
 
-This structure enables the receiver to easily manage incoming requests by iterating over all requests with a given subspace ID and receiver address, which will be the most frequently used query.
+This structure enables Desmos to easily manage requests by iterating over all requests with a given subspace ID and post ID when a post is moved or deleted.
```
Comment (dadamu) [inline docs/architecture/adr-020-post-ownership-change.md:74]: @RiccardoM 
I have identified an issue with this design: we are unable to determine the post's ownership receiver when it is deleted or moved. This leads to a time complexity of O(n) when deleting a post, as it requires the following steps:
- Iterate over all requests by subspace ID.
- Find the post ID of the request that matches the post to be deleted.
- Delete the corresponding request store key.

To address this issue, I have simplified the store key implementation, resulting in an improved time complexity of O(1) for the post deletion process.

File: app/params/weights.go [hunk 2]
```diff
@@ -37,13 +37,6 @@ const (
 	DefaultWeightMsgGrantAllowance              int = 20
 	DefaultWeightMsgRevokeAllowance             int = 5
 
-	DefaultWeightMsgCreatePost           int = 80
```
Comment (dadamu) [inline app/params/weights.go:40]: Move to `operation.go` to be easier management

File: docs/architecture/adr-020-post-ownership-change.md [hunk 3]
```diff
@@ -202,8 +202,8 @@ message QueryIncomingPostOwnerTransferRequestsResponse {
 
 ### Backwards Compatibility
 
-The solution outlined above is **not** backwards compatible and will require a migration script to update all existing posts to the new version. This script will handle the following tasks:
-- migrate all posts to have a new __owner__ field.
+The solution outlined above is fully backwards compatible since it introduces a new __owner__ field for post, but it will require a migration script to update all existing posts. This script will handle the following tasks:
+- migrate all posts to set a new __owner__ field to its __author__.
```
Comment (dadamu) [inline docs/architecture/adr-020-post-ownership-change.md:206]: Adding a new field to proto is fully backwards compatible

**Comment by RiccardoM** [comment]
@dadamu Can you rebase this on the `master` branch please?

**Comment by dadamu** [comment]
@RiccardoM Rebased

File: proto/desmos/posts/v3/msgs.proto [hunk 4]
```diff
@@ -354,4 +377,130 @@ message MsgMovePostResponse {
...
+
+// MsgCancelPostOwnerTransferResponse defines the Msg/CancelPostOwnerTransfer response type
+//
+// Since: Desmos 6.0.0
+message MsgCancelPostOwnerTransferResponse {}
+
+// MsgAcceptPostOwnerTransfer represents a message used to accept a incoming post transfer request
+//
+// Since: Desmos 6.0.0
+message MsgAcceptPostOwnerTransfer {
```
Comment (RiccardoM) [inline proto/desmos/posts/v3/msgs.proto:451]: I think it's better to name them `Msg[Accept/Cancel/Refuse]PostOwnerTransferRequest`, similar to the `Msg[Accept/Cancel/Refuse]DTagTransferRequest` we have

File: x/posts/keeper/msg_server_post_owner_transfer.go [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+// CancelPostOwnerTransfer defines the rpc method for Msg/CancelPostOwnerTransfer
+func (k msgServer) CancelPostOwnerTransfer(goCtx context.Context, msg *types.MsgCancelPostOwnerTransfer) (*types.MsgCancelPostOwnerTransferResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// Check if the sender has profile
+	if !k.HasProfile(ctx, msg.Sender) {
+		return nil, errors.Wrapf(sdkerrors.ErrInvalidRequest, "you cannot cancel a post owner transfer request without having a profile")
+	}
```
Comment (RiccardoM) [inline x/posts/keeper/msg_server_post_owner_transfer.go:82]: Is this wanted? If the request was created, it means that the user had the profile at that time. I think we should allow them to cancel the request even if they have deleted them in the meanwhile

Comment (dadamu) [reply x/posts/keeper/msg_server_post_owner_transfer.go]: Yeah, we should allow all the users to cancel/refuse requests. Updated.

File: proto/desmos/posts/v3/msgs.proto (hunk 4 above)
Comment (dadamu) [reply proto/desmos/posts/v3/msgs.proto]: Updated

File: x/posts/keeper/msg_server_post_owner_trasnfer_test.go [hunk 6]
```diff
@@ -0,0 +1,803 @@
...
+			msg: types.NewMsgRequestPostOwnerTransfer(
+				1,
+				1,
+				"cosmos13t6y2nnugtshwuy0zkrq287a95lyy8vzleaxmd",
+				"cosmos1eqpa6mv2jgevukaqtjmx5535vhc3mm3cf458zg",
+			),
+			shouldErr: true,
+		},
+		{
+			name: "request has already existed returns error",
```
Comment (RiccardoM) [inline x/posts/keeper/msg_server_post_owner_trasnfer_test.go:160]:
```suggestion
			name: "request already exists returns error",
```

File: x/posts/keeper/msg_server_post_owner_trasnfer_test.go [hunk 7]
```diff
@@ -0,0 +1,803 @@
...
+			msg: types.NewMsgRequestPostOwnerTransfer(
+				1,
+				1,
+				"cosmos13t6y2nnugtshwuy0zkrq287a95lyy8vzleaxmd",
+				"cosmos1eqpa6mv2jgevukaqtjmx5535vhc3mm3cf458zg",
+			),
+			shouldErr: true,
+		},
+		{
+			name: "correct data is performed properly",
```
Comment (RiccardoM) [inline x/posts/keeper/msg_server_post_owner_trasnfer_test.go:213]:
```suggestion
			name: "valid transfer is performed properly",
```

File: x/posts/keeper/msg_server_post_owner_trasnfer_test.go [hunk 8]
```diff
@@ -0,0 +1,803 @@
...
+			},
+			msg: types.NewMsgCancelPostOwnerTransferRequest(
+				1,
+				1,
+				"cosmos1eqpa6mv2jgevukaqtjmx5535vhc3mm3cf458zg",
+			),
+			shouldErr: true,
+		},
+		{
+			name: "correct data is performed properly",
```
Comment (RiccardoM) [inline x/posts/keeper/msg_server_post_owner_trasnfer_test.go:368]:
```suggestion
			name: "valid request is performed properly",
```

File: x/posts/keeper/msg_server_post_owner_trasnfer_test.go [hunk 9]
```diff
@@ -0,0 +1,803 @@
...
+			},
+			msg: types.NewMsgRefusePostOwnerTransferRequest(
+				1,
+				1,
+				"cosmos13t6y2nnugtshwuy0zkrq287a95lyy8vzleaxmd",
+			),
+			shouldErr: true,
+		},
+		{
+			name: "correct data is performed properly",
```
Comment (RiccardoM) [inline x/posts/keeper/msg_server_post_owner_trasnfer_test.go:735]:
```suggestion
			name: "correct request is performed properly",
```

File: x/posts/keeper/msg_server_post_owner_trasnfer_test.go [hunk 10]
```diff
@@ -0,0 +1,803 @@
...
+			},
+			msg: types.NewMsgAcceptPostOwnerTransferRequest(
+				1,
+				1,
+				"cosmos13t6y2nnugtshwuy0zkrq287a95lyy8vzleaxmd",
+			),
+			shouldErr: true,
+		},
+		{
+			name: "correct data is performed properly",
```
Comment (RiccardoM) [inline x/posts/keeper/msg_server_post_owner_trasnfer_test.go:599]:
```suggestion
			name: "correct request is performed properly",
```

**Review by RiccardoM — APPROVED** [review]
Great job :100:

**Pull request merged**